### PR TITLE
UPSTREAM: 33958: add global timeout flag

### PR DESCRIPTION
**Pull request by juanvallejo**
UPSTREAM: https://github.com/kubernetes/kubernetes/pull/33958

Related Trello card:
https://trello.com/c/5YkspaNk/711-5-oc-global-timeout-ux-p4

Related Bugzilla RFE: https://bugzilla.redhat.com/show_bug.cgi?id=1358393

This patch adds a global timeout flag (viewable with `oc options`) with
a default value of `0s` (meaning no timeout). It renames all local `timeout` flags
previously declared in specific commands, such as `oc delete` and `oc
scale`, to avoid conflict with the global timeout flag.

**TODO**

- [x] Update [request method](https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/pkg/client/restclient/request.go#L797) to a timeout-based approach

**To Resolve**

- [x] Not sure what to do with commands that already implement a local `--timeout` flag with custom values / descriptions.

**Example**
```
$ oc whoami # no timeout flag set - default to 0s (which means no timeout)
system:admin

$ oc whoami --request-timeout=0 # zero means no timeout
system:admin

$ oc whoami --request-timeout=1ms
Unable to connect to the server: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
```

cc @openshift/cli-review @smarterclayton @jwforres @deads2k @enj

### CODE REVIEW

File: vendor/k8s.io/kubernetes/pkg/kubectl/cmd/rollingupdate.go [hunk 1]
```diff
@@ -95,7 +94,6 @@ func NewCmdRollingUpdate(f *cmdutil.Factory, out io.Writer) *cobra.Command {
 	}
 	cmd.Flags().Duration("update-period", updatePeriod, `Time to wait between updating pods. Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".`)
 	cmd.Flags().Duration("poll-interval", pollInterval, `Time delay between polling for replication controller status after the update. Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".`)
-	cmd.Flags().Duration("timeout", timeout, `Max time to wait for a replication controller to update before giving up. Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".`)
```
Comment (juanvallejo) [inline vendor/k8s.io/kubernetes/pkg/kubectl/cmd/rollingupdate.go]: @fabianofranz I am unsure how we would handle cases where timeout has a specific description (or default value) for a command. For example, [newtoken.go][(https://github.com/openshift/origin/blob/master/pkg/cmd/cli/sa/newtoken.go#L87) has a timeout of `30s`.

**Comment by smarterclayton** [comment]
Timeout 10s is too short for a default.  I may be missing context, but we
explicitly don't have a timeout in rhc for this reason - it may not be
appropriate to set a default timeout.

**Comment by juanvallejo** [comment]
@smarterclayton 
> Timeout 10s is too short for a default.  I may be missing context, but we
> explicitly don't have a timeout in rhc for this reason - it may not be
> appropriate to set a default timeout.

Okay, this makes sense, I'll set the starting value at `0`, which would imply an infinite amount of time before timing out.

I was hoping I could get your feedback on how to approach the actual implementation of timing out requests. I'm looking at [k8s.io/kubernetes/pkg/client/restclient/request.go#L797](https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/pkg/client/restclient/request.go#L797) where the current implementation does not currently have a time limit per request, but rather a fixed number of max retries (10 at the moment) and a small timeout in between each, set by the `Retry-After` header. I was thinking of keeping this implementation exactly as it is, and then adding a "timer" that ends the current request (and ignores any remaining retries) after the `global timeout` time is exceeded. WDYT?

cc @fabianofranz

**Comment by liggitt** [comment]
what does the timeout apply to? dialing? time to first byte? time without data being read? total time?

**Comment by juanvallejo** [comment]
@liggitt 
> what does the timeout apply to? dialing? time to first byte? time without data being read? total time?

I was originally thinking of it applying to total time. Would it make more sense for it to apply to something else? Maybe the timeout applying to dialing would work better, then once it successfully connects to the server, the timeout "timer" would no longer be in effect.

File: vendor/k8s.io/kubernetes/pkg/client/restclient/request.go [hunk 2]
```diff
@@ -793,70 +793,114 @@ func (r *Request) request(fn func(*http.Request, *http.Response)) error {
...
+				return
+			}
+		}
+	}()
 
-				glog.V(4).Infof("Got a Retry-After %s response for attempt %d to %v", seconds, retries, url)
-				r.backoffMgr.Sleep(time.Duration(seconds) * time.Second)
-				return false
+	// ignore timeout if timeout value is zero
+	if r.timeout == 0 {
```
Comment (juanvallejo) [inline vendor/k8s.io/kubernetes/pkg/client/restclient/request.go]: PTAL at this implementation, it works as expected, however maybe there is a cleaner way to "time out" requests?

**Comment by fabianofranz** [comment]
This timeout applies to total time, the same way the higher-level `Timeout` field of `http.Client` [works](https://golang.org/pkg/net/http/#Client), which is what users perceives as "timeout" as in the time it takes for me to get a response.

About defaulting, in case a timeout was not explicitly provided then you just don't set it (not even to `0`, forever, etc) and let it use whatever the API defaults to, which is what happens today AFAIK. I believe the defaults are what's in `DefaultTransport` [here](https://golang.org/pkg/net/http/#RoundTripper), with different values specifically set for dialing, handshaking, etc.

**Comment by fabianofranz** [comment]
Also make sure the timeouts work correctly when you are under a proxy with `HTTP_PROXY` and `HTTPS_PROXY` set. You may need to tweak the round tripper in `pkg/util/httpstream/spdy`. Same for commands that do streaming, http upgrade (`kubectl exec` for example).

**Comment by liggitt** [comment]
I'm really unsure about this... many command line commands are made up of multiple API requests, but this is globally hooking the timeout to the client

**Comment by enj** [comment]
Based on the implementation, the timeout is essentially from when the user hits `Enter` to when they get some output on the CLI (this is not technically not the case but it is close).  Thus if a user sets it to `10 seconds`, all requests (x10 for the hard-coded retry logic) together should take less than that timeout.  Thus the timeout in `http.Client` will usually not be the one that timesout; it would be the external one in the `select`.  The `http.Client` is basically being used to curry the timeout value through the code (this probably deserves a harder look).

**Comment by juanvallejo** [comment]
Another idea would be to remove the `select` timeout, but still set the timeout in the `http.Client`. That way requests still timeout, should a user want them to, without ignoring the `10` request retries. Although it would no longer be a "total time" timeout in this case, it would still allow for a consistent maximum time when making requests.

**Comment by juanvallejo** [comment]
Current working example:

```
$ oc whoami # no timeout flag set - default to 0s (which means no timeout)
system:admin

$ oc whoami --timeout=0 # zero means no timeout
system:admin

$oc whoami --timeout=1ms
Unable to connect to the server: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
```

File: vendor/k8s.io/kubernetes/pkg/client/restclient/config.go [hunk 3]
```diff
@@ -185,6 +189,9 @@ func RESTClientFor(config *Config) (*RESTClient, error) {
 	var httpClient *http.Client
 	if transport != http.DefaultTransport {
 		httpClient = &http.Client{Transport: transport}
+		if config.Timeout != 0 {
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/pkg/client/restclient/config.go]: Why do you allow negatives?

File: vendor/k8s.io/kubernetes/pkg/kubectl/cmd/scale.go [hunk 4]
```diff
@@ -89,7 +89,6 @@ func NewCmdScale(f *cmdutil.Factory, out io.Writer) *cobra.Command {
 	cmd.Flags().Int("current-replicas", -1, "Precondition for current size. Requires that the current size of the resource match this value in order to scale.")
 	cmd.Flags().Int("replicas", -1, "The new desired number of replicas. Required.")
 	cmd.MarkFlagRequired("replicas")
-	cmd.Flags().Duration("timeout", 0, "The length of time to wait before giving up on a scale operation, zero means don't wait. Any other values should contain a corresponding time unit (e.g. 1s, 2m, 3h).")
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/pkg/kubectl/cmd/scale.go]: This timeout means something different.

File: vendor/k8s.io/kubernetes/pkg/kubectl/cmd/delete.go [hunk 5]
```diff
@@ -110,9 +110,9 @@ func NewCmdDelete(f *cmdutil.Factory, out io.Writer) *cobra.Command {
 	cmd.Flags().Bool("cascade", true, "If true, cascade the deletion of the resources managed by this resource (e.g. Pods created by a ReplicationController).  Default true.")
 	cmd.Flags().Int("grace-period", -1, "Period of time in seconds given to the resource to terminate gracefully. Ignored if negative.")
 	cmd.Flags().Bool("now", false, "If true, resources are force terminated without graceful deletion (same as --grace-period=0).")
-	cmd.Flags().Duration("timeout", 0, "The length of time to wait before giving up on a delete, zero means determine a timeout from the size of the object")
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/pkg/kubectl/cmd/delete.go]: This timeout means something different

Comment (juanvallejo) [inline vendor/k8s.io/kubernetes/pkg/kubectl/cmd/delete.go]: would it be okay to rename these timeouts to something like `--delete-timeout` and `--scale-timeout`, or would it make more sense to rename the global timeout to `--global-timeout`?

File: vendor/k8s.io/kubernetes/pkg/client/restclient/config.go (hunk 3 above)
Comment (juanvallejo) [inline vendor/k8s.io/kubernetes/pkg/client/restclient/config.go]: Updated this here: https://github.com/openshift/origin/pull/11104/commits/4a7769ff0233bb104041c3dc8430f14a4fb12c22

File: vendor/k8s.io/kubernetes/pkg/kubectl/cmd/scale.go (hunk 4 above)
Comment (juanvallejo) [inline vendor/k8s.io/kubernetes/pkg/kubectl/cmd/scale.go]: Addressed in https://github.com/openshift/origin/pull/11104/commits/faf8b4a806370aaefd78170df70c9654f65ddd62

**Comment by deads2k** [comment]
> would it be okay to rename these timeouts to something like --delete-timeout and --scale-timeout, or would it make more sense to rename the global timeout to --global-timeout?

Renaming flags is a pretty big deal for scripts.  That's why we have @openshift/api-review  take a look at  new flags.  `--request-timeout` maybe?

**Comment by juanvallejo** [comment]
@deads2k 
> Renaming flags is a pretty big deal for scripts. That's why we have @openshift/api-review take a look at new flags. --request-timeout maybe?

Thanks for the feedback, I have updated the global flag's name to `--request-timeout`, I will do the same for the upstream PR.

**Comment by juanvallejo** [comment]
[test]

**Comment by juanvallejo** [comment]
check, integration, conformance test flaked on https://github.com/openshift/origin/issues/11292 re[test]

**Comment by juanvallejo** [comment]
check, integration, conformance tests flaked on #11292 re[test]

**Comment by juanvallejo** [comment]
[integration](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin_integration/6909/) test flaked on https://github.com/openshift/origin/issues/11240 re[test]

**Comment by juanvallejo** [comment]
[integration](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin_integration/6925/) flaked on https://github.com/openshift/origin/issues/11240 re[test]

**Comment by juanvallejo** [comment]
https://github.com/openshift/origin/issues/8571 re[test]

**Comment by juanvallejo** [comment]
https://github.com/openshift/origin/issues/8571 re[test]

**Comment by juanvallejo** [comment]
integration flaked on https://github.com/openshift/origin/issues/11292 re[test]

**Comment by juanvallejo** [comment]
@fabianofranz upstream merged! https://github.com/kubernetes/kubernetes/pull/33958 PTAL

**Comment by fabianofranz** [comment]
LGTM. Hold on to your seats, [merge].

**Comment by fabianofranz** [comment]
@juanvallejo failed on outdated completions, please update and I'll tag again.

**Comment by juanvallejo** [comment]
@fabianofranz Done! rebased with latest master and updated docs / completions

**Comment by fabianofranz** [comment]
[merge]

**Pull request merged**